pkg/metaserver/metarequest.go

```diff
@@ -0,0 +1,104 @@
+package metaserver
+
+import (
+	"net/http"
+	"strings"
+
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	apirequest "k8s.io/apiserver/pkg/endpoints/request"
+)
+
+const (
+	MetaGroup           = "meta"
+	MetaVersion         = "v1"
+	resourceListNameSep = "---"
+)
+
+var (
+	MetaGroupVersion = schema.GroupVersion{Group: MetaGroup, Version: MetaVersion}
+	APIVersions      = MetaGroupVersion.WithResource("apiversions")
+	APIGroupLists    = MetaGroupVersion.WithResource("apigrouplists")
+	APIGroups        = MetaGroupVersion.WithResource("apigroups")
+	APIResourceLists = MetaGroupVersion.WithResource("apiresourcelists")
+)
+
+// DecorateMetaRequest check if the request is meta request and transform it to
+// get request.
+// meta request is that wants to get resource which is belong to Group meta like:
+// Path				NewGetPath										Want Resources
+// /api				/api/meta/v1/apiversions/cloud-apiversions		metav1.APIVersions
+// /apis			/api/meta/v1/apigroups/cloud-apigroups			metav1.APIGroups
+// /api/v1			/api/meta/v1/apiresourcelists/{SEP}v1			metav1.APIResourcesList in /v1
+// /apis/apps/v1	/api/meta/v1/apiresourcelists/apps{SEP}v1		metav1.APIResourcesList in apps/v1
+func DecorateMetaRequest(handler http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+		ctx := req.Context()
+		reqInfo, ok := apirequest.RequestInfoFrom(ctx)
+		if ok {
+			if gvr, name, ok := IsMetaRequest(reqInfo); ok {
+				// transform to normal get request
+				reqInfo.IsResourceRequest = true
+				reqInfo.Verb = "get"
+				reqInfo.APIPrefix = "apis"
+				reqInfo.APIGroup = gvr.Group
+				reqInfo.APIVersion = gvr.Version
+				reqInfo.Resource = gvr.Resource
+				reqInfo.Path = "/apis/meta/v1/" + gvr.Resource + name
+				reqInfo.Name = name
+				req = req.WithContext(apirequest.WithRequestInfo(req.Context(), reqInfo))
+			}
+		}
+		handler.ServeHTTP(w, req)
+	})
+}
+
+func IsMetaRequest(info *apirequest.RequestInfo) (gvr schema.GroupVersionResource, name string, ok bool) {
+	switch {
+	case info.Path == "/api":
+		return APIVersions, "cloud-apiversions", true
+	case info.Path == "/apis":
+		return APIGroupLists, "cloud-apigrouplist", true
+	case info.Path == "/api/v1":
+		gv := schema.GroupVersion{Group: "", Version: "v1"}
+		name := ResourceListNameParser.New(gv)
+		return APIResourceLists, name, true
+	case strings.HasPrefix(info.Path, "/apis"):
+		path := strings.TrimRight(info.Path, "/")
+		slice := strings.Split(path, "/")
+		switch len(slice) {
+		case 3: // /apis/group
+			groupIndex := 2
+			return APIGroups, slice[groupIndex], true
+		case 4: // /apis/group/version
+			groupIndex := 2
+			versionIndex := 3
+			gv := schema.GroupVersion{Group: slice[groupIndex], Version: slice[versionIndex]}
+			name := ResourceListNameParser.New(gv)
+			return APIResourceLists, name, true
+		}
+	}
```

Comment:
I would use this style of code
```go
        switch info.Path {
        case "/api":
                return APIVersions, "cloud-apiversions", true
        case "/apis":
                return APIGroupLists, "cloud-apigrouplist", true
        case "/api/v1":
                gv := schema.GroupVersion{Group: "", Version: "v1"}
                return APIResourceLists, ResourceListNameParser.New(gv), true
        }

        if strings.HasPrefix(info.Path, "/apis") {
                path := strings.TrimRight(info.Path, "/")
                slice := strings.Split(path, "/")
                switch len(slice) {
                case 3: // /apis/group
                        groupIndex := 2
                        return APIGroups, slice[groupIndex], true
                case 4: // /apis/group/version
                        groupIndex, versionIndex := 2, 3
                        gv := schema.GroupVersion{Group: slice[groupIndex], Version: slice[versionIndex]}
                        return APIResourceLists, ResourceListNameParser.New(gv), true
                }
        }
        return schema.GroupVersionResource{}, "", false
```
The function already specifies that the second return value is `name`, so it may not be necessary to declare a new `name` to indicate what the `ResourceListNameParser.New(gv)` means

Reply:
@Iceber Seems Good! Thx